### PR TITLE
Adding built-in sorting capability to _cat apis.

### LINKED ISSUE #16975: Built-in sorting in _cat APIs
While using `sort(1)` is great and all, more often than not I find myself using browser for `_cat` APIs. It's easy to bookmark and reuse.

Unfortunately, `/_cat/indices?v` and similar endpoints have undefined order of rows and it is not always convenient. Moreover, `?v` makes using `sort(1)` less convenient, since header is sorted too.

I suggest adding another parameter to sort on any column list. It would accept the same thing as `h`. Example request to sort on index time:

```
curl -s 'http://127.0.0.1:9200/_cat/indices?v&h=index,pri,docs,indexing.index_time&s=indexing.index_time'
```

### PR DESCRIPTION
**Pull request by a2lin**
This PR adds a parameter to _cat apis that causes the rows returned by the api invocation to be sorted by a set of provided columns (header names).

Sorting on any object is supported, with .compareTo() being invoked for comparable objects, nulls being treated as the smallest object, and .toString() called on objects not implementing comparable before comparing them. Multiple header names can be provided, and they are applied to the sort in the order that they were provided.

An example invocation would be: `localhost:9200/_cat/templates?s=name,template`.
Aliases are also supported, so `localhost:9200/_cat/templates?s=n,t` would have an identical result to the first invocation. 

Sorting is by default ascending, and can be made descending on a per-column basis by adding :desc to the header/header alias name. For example, `localhost:9200/_cat/templates?s=n:desc,t`.

Also included in this PR are compareTo methods for ByteSizeValue, SizeValue, and TimeValue, as these are the three 'values' that are special cased in RestTable.

Examples:

Simple example (sorting by name):

```
curl "localhost:9200/_cat/templates?v&s=name"
name                  template     order version
pizza_hawaiian        *pineapples* 1     
pizza_pepperoni       *pepperoni*  1     
sushi_california_roll *avocado*    1     1
```

Alias example (name sort with alias):

```
curl "localhost:9200/_cat/templates?v&s=n"
name                  template     order version
pizza_hawaiian        *pineapples* 1     
pizza_pepperoni       *pepperoni*  1     
sushi_california_roll *avocado*    1     1
```

Reverse example (inverse name sort):

```
curl "localhost:9200/_cat/templates?v&s=name:desc"
name                  template     order version
sushi_california_roll *avocado*    1     1
pizza_pepperoni       *pepperoni*  1     
pizza_hawaiian        *pineapples* 1     
```

Reverse alias example (inverse name sort with alias):

```
curl "localhost:9200/_cat/templates?v&s=n:desc"
name                  template     order version
sushi_california_roll *avocado*    1     1
pizza_pepperoni       *pepperoni*  1     
pizza_hawaiian        *pineapples* 1       
```

Null example (sort by version, some version values are missing):

```
curl "localhost:9200/_cat/templates?v&s=version"
name                  template     order version
pizza_hawaiian        *pineapples* 1     
pizza_pepperoni       *pepperoni*  1     
sushi_california_roll *avocado*    1     1
```

Multisort example (first sort by version, then by template):

```
curl "localhost:9200/_cat/templates?v&s=version,template"
name                  template     order version
pizza_pepperoni       *pepperoni*  1     
pizza_hawaiian        *pineapples* 1     
sushi_california_roll *avocado*    1     1
```

Multisort alias example (version desc alias, then template)

```
curl "localhost:9200/_cat/templates?v&s=v:desc,t"
name                  template     order version
sushi_california_roll *avocado*    1     1
pizza_pepperoni       *pepperoni*  1     
pizza_hawaiian        *pineapples* 1      
```

Sorting bytes example:

```
curl "localhost:9200/_cat/indices?v&h=health,status,index,docs.count,store.size&s=store.size"
health status index    docs.count store.size
yellow open   twix              0       795b
yellow open   twizzler          2      3.3kb
yellow open   tweeter          10      6.9kb
```

Inverted sorting bytes example:

```
curl "localhost:9200/_cat/indices?v&h=health,status,index,docs.count,store.size&s=store.size:desc"
health status index    docs.count store.size
yellow open   tweeter          10      6.9kb
yellow open   twizzler          2      3.3kb
yellow open   twix              0       795b
```

Sorting numbers example:

```
curl "localhost:9200/_cat/indices?v&h=health,status,index,docs.count,store.size&s=docs.count:desc"
health status index    docs.count store.size
yellow open   tweeter          10      6.9kb
yellow open   twizzler          2      3.3kb
yellow open   twix              0       795b
```

Closes #16975

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: What about just converting to bytes and comparing? The way you have it now this isn't consistent with `equals`....

Also the _cat API we call `toString` which doesn't really use the unit anyway.

File: core/src/main/java/org/elasticsearch/common/unit/SizeValue.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/unit/SizeValue.java]: Same comment as the last class - I'd just convert to `SINGLE` here.

The equals method is funny on this class - I'd change it so it lines converts to `SINGLE` as well.

File: core/src/main/java/org/elasticsearch/common/unit/TimeValue.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/unit/TimeValue.java]: Same deal, I'd just convert it to nanos.

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java]: nit: we usually add a space here. We don't have anything that enforces this style but we usually do.

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java]: Maybe return an `int[]` so this doesn't have to box?

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java]: I'd do something like
```
boolean reverse = false;
if (columnOrdering[i].endsWith(":desc")) {
  columnOrder[i] = columnOrder[i].substring(...);
} else if (columnOrdering[i].endsWith(":asc")) {
  columnOrder[i] = columnOrder[i].substring(...);
}
```

or something like that. I think we should complain if it ends in something other than `:desc` and `:asc`.

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java]: Ah, you'd need to box anyway to do the sorting so far as I know so this isn't worth it I think.

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java]: I think we should complain if we don't find the header name.

File: docs/reference/cat.asciidoc
Comment (nik9000) [inline docs/reference/cat.asciidoc]: Can you add this like this:

```
[source,sh]
--------------------------------------------------
_cat/templates?v&s=order:desc,template
--------------------------------------------------
// CONSOLE

returns:

[source,sh]
--------------------------------------------------
name                  template     order version
pizza_pepperoni       *pepperoni*  2
sushi_california_roll *avocado*    1     1
pizza_hawaiian        *pineapples* 1
--------------------------------------------------
```

Technically you can add `// TESTRESPONSE` to the bottom of the last snippet and we'd test that the response matches during the build (`gradle docs:check`).

**Review by nik9000 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java
Comment (a2lin) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: I had it like that originally, but changed it since if we convert to Bytes and compare, every value above (Long.MAX_LONG, BYTES) will be treated as equals. 

Is there a reason why we want this behavior? (I noticed the writeTo method also writes getBytes())

For the _cat api, I think it uses ByteSizeValue directly in some cases to support the ability to change numeric formats, e.g. the ?bytes=b thing.

**Comment by a2lin** [comment]
@nik9000 Thanks so much for reviewing this PR! 

I've implemented the requested changes, although had a question about the value type comparisons that I noted above.

**Comment by nik9000** [comment]
> I've implemented the requested changes, although had a question about the value type comparisons that I noted above.

Sorry I didn't get back to you earlier! I'm stuck in meetings all day every day for the next few days and don't really have much energy for reading emails at the end of the day but I'll review again now!

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java]: ++

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: > Is there a reason why we want this behavior? (I noticed the writeTo method also writes getBytes())

I'm not sure why we wanted it at some point, but we've gone away from that for `TimeValue` so I expect we should go away from it for `ByteSizeValue` eventually. I think maybe in another PR though rather than making this one bigger.

Yeah, the max bytes problem makes sense to me. I guess we should just make the equals and compareTo actually correct rather than have the rollover issue. I don't think we need equals or compareTo to be super efficient so I'd prefer if we just made the right even if it isn't fast. I suspect we've never hit this because `Long.MAX_VALUE` is so much bigger than anything anyone is particularly likely to provide as input. But, yeah, I'd love it if you made the comparator work properly for these values anyway.

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java]: I'm not sure the length comparison buys you much. I'm fairly sure `String#endsWith` already checks that super duper early.

File: core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java]: Rather than `- 5` I'd do `- ":desc".length()` just to make it super clear what is up.

File: core/src/test/java/org/elasticsearch/rest/action/cat/RestTableTests.java
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/rest/action/cat/RestTableTests.java]: I think it'd be nice to do:
```java
Exception e = expectThrows(UnsupportedOperationException.class, () -> {
  RestTable.getRowOrder(table, restRequest);
});
assertEquals("blah blah blah", e.getMessage());
```

It is just nice to assert something about the message so you know *which* `UnsupportedOperationException` you get.

**Review by nik9000 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java
Comment (a2lin) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: Should I keep my original behavior and make equals defer to the new compareTo? 

(The compareTo that was implemented tries to find the minimum of the two ByteSizeUnits and converts both to them, causing this to work out properly)

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: Yeah, looking at it again, that makes sense!

**Comment by a2lin** [comment]
@nik9000 Thanks again for reviewing! I've changed the value types back to the way they were in the original commit, and made the changes you requested.

**Comment by nik9000** [comment]
@a2lin I just realized I'd forgotten about this! I've re-added it to me review queue and should get to it shortly.

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: Finally getting back to this! On this look I noticed that this would make hashcode and equals inconsistent which is yet another thing to fix. I haven't really thought through the right way to make them consistent, but there ought to be a way!

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: Sorry for all the trouble with this! It was a mess before and you are only making it better and this is really just another thing to do to make it better but hashcode and equals being consistent is one of those cardinal rules of java....

**Review by nik9000 — CHANGES_REQUESTED** [review]

**Comment by a2lin** [comment]
@nik9000 Sorry for the delay in responding!

I changed it back to your originally-suggested solution (converting to bytes, singles, nanos), but fixed the edge problems by casting to double. I think this solves the compareTo, Equals, HashCode, issues.

Thanks again for the reviews!

File: core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java
Comment (a2lin) [inline core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java]: Heh. I think I failed pretty hard on this one but this new solution seems a bit better?

File: core/src/main/java/org/elasticsearch/common/unit/TimeValue.java
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/unit/TimeValue.java]: use `Double.compareTo`

**Comment by a2lin** [comment]
@s1monw I've fixed this. Thanks for pointing it out to me!

**Review by nik9000 — APPROVED** [review]

**Comment by nik9000** [comment]
I think that should do it! I'll test locally soon and see about merging! Thanks again for this.

**Comment by nik9000** [comment]
I've attached `v5.1.0` and `v6.0.0` labels. I'll attempt to merge it to get into those releases.

**Comment by nik9000** [comment]
Ok! All passing locally. It requires a one line change to make it compatible with https://github.com/elastic/elasticsearch/pull/20722 which I'll push after merging your change. Awesome!

**Comment by a2lin** [comment]
@nik9000 Thanks for reviewing and merging!

**Comment by nik9000** [comment]
> @nik9000 Thanks for reviewing and merging!

Thanks for being so persistent with the reviews!

Master: 10ddcc46167ab6ced10dc4fb85638d5d240fce04 (my compatibility patch: f3b8e7c23813863b562ec7399d1268d3a94d7f68)
5.x: 4b5f719692c163f1cf0fc036abbe56ce5a64c32e (my compatibility patch: 504a9cd1c3f9dfa4d42bc7a499ae2b05d6a79b4e)

**Pull request merged**
